pkg/sql/distsqlrun/algebraic_set_op.go

```diff
@@ -88,183 +98,250 @@ func newAlgebraicSetOp(
 }
 
 func (e *algebraicSetOp) Run(wg *sync.WaitGroup) {
+	if e.out.output == nil {
+		panic("algebraicSetOp output not initialized for emitting rows")
+	}
+	Run(e.flowCtx.Ctx, e, e.out.output)
 	if wg != nil {
-		defer wg.Done()
+		wg.Done()
 	}
+}
 
-	ctx := log.WithLogTag(e.flowCtx.Ctx, "ExceptAll", nil)
-	ctx, span := processorSpan(ctx, "exceptAll")
-	defer tracing.FinishSpan(span)
-
-	log.VEventf(ctx, 2, "starting exceptAll set process")
-	defer log.VEventf(ctx, 2, "exiting exceptAll")
+// producerMeta constructs the ProducerMetadata after consumption of rows has
+// terminated, either due to being indicated by the consumer, or because the
+// processor ran out of rows or encountered an error. It is ok for err to be
+// nil indicating that we're done producing rows even though no error occurred.
+func (e *algebraicSetOp) producerMeta(err error) *ProducerMetadata {
+	var meta *ProducerMetadata
+	if !e.closed {
+		if err != nil {
+			meta = &ProducerMetadata{Err: err}
+		} else if trace := getTraceData(e.ctx); trace != nil {
+			meta = &ProducerMetadata{TraceData: trace}
+		}
+		// We need to close as soon as we send producer metadata as we're done
+		// sending rows. The consumer is allowed to not call ConsumerDone().
+		e.internalClose()
+	}
+	return meta
+}
 
-	defer e.leftSource.ConsumerDone()
-	defer e.rightSource.ConsumerDone()
+func (e *algebraicSetOp) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
+	var row sqlbase.EncDatumRow
+	var meta *ProducerMetadata
 
-	e.evalCtx = e.flowCtx.NewEvalCtx()
+	for {
+		switch e.opType {
+		case AlgebraicSetOpSpec_Except_all:
+			row, meta = e.nextExceptAll()
```

Comment:
It seems to me that this is a relatively easy opportunity to static-ify the responsibility of this processor. Can we define an interface for algebraic set ops (doesn't have to be a Go interface, just like a protocol) and then plug the correct one into this processor at planning time? That way, we won't need to perform this switch on every row. I'm not trying to suggest you do that in this PR- just noticed this and writing it down.